src/cregistry/distfile.c

```diff
@@ -0,0 +1,316 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include "distfile.h"
+#include "util.h"
+#include "registry.h"
+#include "sql.h"
+
+#include <sqlite3.h>
+#include <stdlib.h>
+#include <string.h>
+
+/**
+ * Converts a `sqlite3_stmt` into a `reg_distfile`. The first column of the stmt's
+ * row must be the id of a distfile; the second column must be the path of a distfile;
+ * the third either `SQLITE_NULL` or the address of the entry in memory.
+ *
+ * @param [in] userdata sqlite3 database
+ * @param [out] distfile    distfile described by `stmt`
+ * @param [in] stmt     `sqlite3_stmt` with appropriate columns
+ * @param [out] errPtr  unused
+ * @return              true if success; false if failure
+ */
+static int reg_stmt_to_distfile(void* userdata, void** distfile, void* stmt,
+        void* calldata UNUSED, reg_error* errPtr UNUSED) {
+    int is_new;
+    reg_registry* reg = (reg_registry*)userdata;
+    reg_distfile_pk key;
+    Tcl_HashEntry* hash;
+    char* hashkey;
+
+    key.id = sqlite3_column_int64(stmt, 0);
+    key.subdir = strdup((const char*) sqlite3_column_text(stmt, 1));
+    if (!key.subdir) {
+        return 0;
+    }
+    key.path = strdup((const char*) sqlite3_column_text(stmt, 2));
+    if (!key.path) {
+        free(key.subdir);
+        return 0;
+    }
+
+    hashkey = sqlite3_mprintf("%lld:%s", key.id, key.path);
+    if (!hashkey) {
+        free(key.subdir);
+        free(key.path);
+        return 0;
+    }
+    hash = Tcl_CreateHashEntry(&reg->open_distfiles,
+            hashkey, &is_new);
+    sqlite3_free(hashkey);
+
+    if (is_new) {
+        reg_distfile* f = malloc(sizeof(reg_distfile));
+        if (!f) {
+            free(key.subdir);
+            free(key.path);
+            return 0;
```

Comment:
Should we also add a Tcl_DeleteHashEntry() here, since the memory for the new entry has already been allocated? This is probably missing in all the other implementations, too :(